hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotPurgeRequest.java

```diff
@@ -80,33 +87,62 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager, TermIn
       Map<String, SnapshotInfo> updatedPathPreviousAndGlobalSnapshots =
           new HashMap<>();
 
-      // Snapshots that are purged by the SnapshotDeletingService
-      // will update the next snapshot so that is can be deep cleaned
-      // by the KeyDeletingService in the next run.
+      // Each snapshot purge operation does three things:
+      //  1. Update the snapshot chain,
+      //  2. Update the deep clean flag for the next active snapshot (So that it can be
+      //     deep cleaned by the KeyDeletingService in the next run),
+      //  3. Finally, purge the snapshot.
+      // All of these steps have to be performed only when it acquires all the necessary
+      // locks (lock on the snapshot to be purged, lock on the next active snapshot, and
+      // lock on the next path and global previous snapshots). Ideally, there is no need
+      // for locks for snapshot purge and can rely on OMStateMachine because OMStateMachine
+      // is going to process each request sequentially.
+      //
+      // But there is a problem with that. After filtering unnecessary SST files for a snapshot,
+      // SstFilteringService updates that snapshot's SstFilter flag. SstFilteringService cannot
+      // use SetSnapshotProperty API because it runs on each OM independently and One OM does
+      // not know if the snapshot has been filtered on the other OM in HA environment.
+      //
+      // If locks are not taken snapshot purge and SstFilteringService will cause a race condition
+      // and override one's update with another.
       for (String snapTableKey : snapshotDbKeys) {
-        SnapshotInfo fromSnapshot = omMetadataManager.getSnapshotInfoTable()
-            .get(snapTableKey);
-
-        if (fromSnapshot == null) {
-          // Snapshot may have been purged in the previous iteration of SnapshotDeletingService.
-          LOG.warn("The snapshot {} is not longer in snapshot table, It maybe removed in the previous " +
-                  "Snapshot purge request.", snapTableKey);
-          continue;
-        }
+        // To acquire all the locks, a set is maintained which is keyed by snapshotTableKey.
+        // snapshotTableKey is nothing but /volumeName/bucketName/snapshotName.
+        // Once all the locks are acquired, it performs the three steps mentioned above and
+        // release all the locks after that.
+        Set<String> lockSet = new HashSet<>();
+        try {
+          acquireLock(lockSet, snapTableKey, omMetadataManager);
```

Comment:
It would avoid deadlock scenarios if code changes further in future. What I am asking is if next SnapTableKey is lesser than this then we should take a lock on that before this.

Reply:
Ah, I see what you meant. If other operations require nested snapshot locks, can cause a deadlock. And it can be avoided if locks are taken in sorted order. But if nested locks are required, those will be in snapshot chain order (which is the current order) rather than snapshotInfoTable order.
I can create a follow-up task if strongly believe locks should be taken based on sorted table key because it would require some refactoring and right now this is the only operation using nested snapshot locks.